cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java

```diff
@@ -51,22 +52,63 @@
+ *
+ * TODO: set state methods are only called from unit tests. They should be removed in favor of using AppMetadataStore.
  */
 public interface Store extends RuntimeStore {
 
   /**
-   * Logs initialization of program run and persists program status to {@link ProgramRunStatus#STARTING}.
+   * Logs initialization of program run and persists program cluster status to
+   * {@link ProgramRunClusterStatus#PROVISIONING}.
    *
    * @param id run id of the program
    * @param startTime start timestamp in seconds
-   * @param twillRunId Twill run id
    * @param runtimeArgs the runtime arguments for this program run
    * @param systemArgs the system arguments for this program run
    * @param sourceId id of the source of program run status, which is proportional to the timestamp of
    *                 when the current program run status is reached
    */
-  void setStart(ProgramRunId id, long startTime, @Nullable String twillRunId,
-                Map<String, String> runtimeArgs, Map<String, String> systemArgs, byte[] sourceId);
+  void setProvisioning(ProgramRunId id, long startTime, Map<String, String> runtimeArgs,
+                       Map<String, String> systemArgs, byte[] sourceId);
+
+  /**
+   * Persists program run cluster status to {@link ProgramRunClusterStatus#PROVISIONED}.
+   *
+   * @param id run id of the program
+   * @param numNodes number of nodes for the cluster
+   * @param sourceId id of the source of program run status, which is proportional to the timestamp of
+   *                 when the current program run status is reached
+   */
+  void setProvisioned(ProgramRunId id, int numNodes, byte[] sourceId);
+
+  /**
+   * Persists program run cluster status to {@link ProgramRunClusterStatus#DEPROVISIONING}.
+   *
+   * @param id run id of the program
+   * @param sourceId id of the source of program run status, which is proportional to the timestamp of
+   *                 when the current program run status is reached
+   */
+  void setDeprovisioning(ProgramRunId id, byte[] sourceId);
+
+  /**
+   * Persists program run cluster status to {@link ProgramRunClusterStatus#DEPROVISIONED}.
+   *
+   * @param id run id of the program
+   * @param sourceId id of the source of program run status, which is proportional to the timestamp of
+   *                 when the current program run status is reached
+   */
+  void setDeprovisioned(ProgramRunId id, byte[] sourceId);
+
+  /**
+   * Logs initialization of program run and persists program status to {@link ProgramRunStatus#STARTING}.
+   *
+   * @param id run id of the program
+   * @param twillRunId Twill run id
+   * @param systemArgs the system arguments for this program run
+   * @param sourceId id of the source of program run status, which is proportional to the timestamp of
+   *                 when the current program run status is reached
+   */
+  void setStart(ProgramRunId id, @Nullable String twillRunId, Map<String, String> systemArgs, byte[] sourceId);
```

Comment:
we dont need `startTime` here?

Reply:
it's set in the provisioning state, which is the new first state.